File: app/src/main/java/com/ssy/ferry/trace/LooperMonitor.java
package com.ssy.ferry.trace;

import android.os.Build;
import android.os.Looper;
import android.os.MessageQueue;
import android.util.Log;
import android.util.Printer;

import java.lang.reflect.Field;
import java.util.HashSet;

public class LooperMonitor implements MessageQueue.IdleHandler {
    private static final String TAG = LooperMonitor.class.getName();
    private static Printer printer;
    private static final HashSet<LooperDispatchListener> listeners = new HashSet<>();

    @Override
    public boolean queueIdle() {
        resetPrinter();
        return true;
    }
    private static final LooperMonitor monitor = new LooperMonitor();
    public LooperMonitor(){
        resetPrinter();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            Looper.getMainLooper().getQueue().addIdleHandler(this);
        } else {
            MessageQueue queue = reflectObject(Looper.getMainLooper(), "mQueue");
            queue.addIdleHandler(this);
        }
    }

    public abstract static class LooperDispatchListener {

        boolean isHasDispatchStart = false;

        boolean isValid() {
            return false;
        }

        void dispatchStart() {
            this.isHasDispatchStart = true;
        }

        void dispatchEnd() {
            this.isHasDispatchStart = false;
        }
    }

    public static void register(LooperDispatchListener listener) {
        synchronized (listeners) {
            listeners.add(listener);
        }
    }
    public static void unregister(LooperDispatchListener listener) {
        if (null == listener) {
            return;
        }
        synchronized (listeners) {
            listeners.remove(listener);
        }
    }
    private static void resetPrinter() {
        Log.e("haha","resetPrinter");
        final Printer originPrinter = reflectObject(Looper.getMainLooper(), "mLogging");
        Looper.getMainLooper().setMessageLogging(printer = new Printer() {
            boolean isHasChecked = false;
            boolean isValid = false;

            @Override
            public void println(String x) {
                if (null != originPrinter) {
                    originPrinter.println(x);
                }

                if (!isHasChecked) {
                    isValid = x.charAt(0) == '>' || x.charAt(0) == '<';
                    isHasChecked = true;
                    if (!isValid) {
                    }
                }

                if (isValid) {
                    dispatch(x.charAt(0) == '>');
                }

            }
        });

    }
    private static void dispatch(boolean isBegin) {

        for (LooperDispatchListener listener : listeners) {
            if (listener.isValid()) {
                if (isBegin) {
                    if (!listener.isHasDispatchStart) {
                        listener.dispatchStart();
                    }
                } else {
                    if (listener.isHasDispatchStart) {
                        listener.dispatchEnd();
                    }
                }
            } else if (!isBegin && listener.isHasDispatchStart) {
                listener.dispatchEnd();
            }
        }

    }
    private static <T> T reflectObject(Object instance, String name) {
        try {
            Field field = instance.getClass().getDeclaredField(name);
            field.setAccessible(true);
            return (T) field.get(instance);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: app/src/main/java/com/ssy/ferry/core/LooperMonitor.kt
package com.ssy.ferry.core

import android.os.Build
import android.os.Looper
import android.os.MessageQueue
import android.util.Printer
import android.os.Build.VERSION.SDK_INT
import android.util.Log
import com.ssy.ferry.listener.LooperDispatchListener
import kotlin.collections.HashSet


/**
 * 2019-10-16
 * @author Mr.S
 *
 * 这个是我们tracer 的发动机，当 idle 的时候就开始运行
 *
 */
class LooperMonitor : MessageQueue.IdleHandler {
    private val TAG = LooperMonitor::class.java.name
    var printer: Printer? = null


    companion object {
        val looperDispatchListeners = HashSet<LooperDispatchListener>()
        open val looperMonitor = LooperMonitor()

        open fun register(looperDispatchListener: LooperDispatchListener) {
            synchronized(looperDispatchListeners) {
                looperDispatchListeners.add(looperDispatchListener)
            }
        }

    }

    override fun queueIdle(): Boolean {
        repalcePrinter()
        return true
    }

    constructor() {
        repalcePrinter()
        if (SDK_INT >= Build.VERSION_CODES.M) {
            Looper.getMainLooper().queue.addIdleHandler(this)
        } else {
            val queue = reflectObject<MessageQueue>(Looper.getMainLooper(), "mQueue")
            queue?.addIdleHandler(this)
        }
    }

    private fun repalcePrinter() {
        val originPrinter = reflectObject<Printer>(Looper.getMainLooper(), "mLogging")

        Looper.getMainLooper().setMessageLogging( object : Printer {
             var isHasChecked = false
             var isValid = false

            override fun println(x: String) {
                originPrinter?.println(x)

                if (!isHasChecked) {
                    isValid = x[0] == '>' || x[0] == '<'
                    isHasChecked = true
                    if (!isValid) {
                    }
                }

                if (isValid) {
                    dispatch(x[0] == '>')
                }
            }
        })
    }

    private fun dispatch(isStart: Boolean) {
        for (item in looperDispatchListeners) {
            if (item.isAlive()) {
                if (isStart) {
                    item.dispatchStart()
                } else {
                    item.dispatchEnd()
                }

            } else {
                if (item.hasDispatchStart) {
                    item.dispatchEnd()
                }
            }
        }

    }


    private fun <T> reflectObject(instance: Any, name: String): T? {
        try {
            val field = instance.javaClass.getDeclaredField(name)
            field.isAccessible = true
            return field.get(instance) as T
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return null
    }
}